#### 1. 印出每個特徵值和紅酒品質的共變數和相關係數

In [36]:
import pandas as pd
import numpy as np


df = pd.read_csv('winequality-red.csv')
table = np.array(df.values)
x = []
for i in range(table.shape[1] - 1):
    x.append(table[:,i])
y = table[:,11]
y_mean = y.mean()
for i in range(table.shape[1] - 1):
    print("---", df.columns[i], "---")
    x_mean = x[i].mean()
    diff = (x[i] - x_mean)*(y - y_mean)
    covar = diff.sum() / table.shape[0]
    print("Covariance:", covar)
    corr = covar / (x[i].std() * y.std())
    print("Correlation:", corr, "\n")

--- fixed acidity ---
Covariance: 0.1743145047268051
Correlation: 0.12405164911322429 

--- volatile acidity ---
Covariance: -0.056440563813922165
Correlation: -0.3905577802640072 

--- citric acid ---
Covariance: 0.035589621562256896
Correlation: 0.22637251431804134 

--- residual sugar ---
Covariance: 0.015625267668465396
Correlation: 0.013731637340066277 

--- chlorides ---
Covariance: -0.0048964807976842935
Correlation: -0.12890655993005268 

--- free sulfur dioxide ---
Covariance: -0.4276394604038405
Correlation: -0.05065605724427635 

--- total sulfur dioxide ---
Covariance: -4.914161876501144
Correlation: -0.18510028892653782 

--- density ---
Covariance: -0.0002664369733897944
Correlation: -0.17491922778334873 

--- pH ---
Covariance: -0.007193320872449596
Correlation: -0.057731391205382156 

--- sulphates ---
Covariance: 0.034391886580144485
Correlation: 0.2513970790692614 

--- alcohol ---
Covariance: 0.4095327325044069
Correlation: 0.476166324001136 



----------------------------------------------------
#### 2. 建立模型並選擇預測資料
觀察上述資料，挑選出四個關聯性較大的欄位，  
分別是 volatile acidity、citric acid、sulphates、alcohol

- volatile acidity

In [56]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

i = 1
print("---", df.columns[i], "---")
X = pd.DataFrame(x[i], columns=[str(df.columns[i])])
y = pd.Series(y)
XTrain, XTest, yTrain, yTest = train_test_split(X, y, test_size=0.3,
                                                random_state=5)
lm = LinearRegression()
lm.fit(XTrain, yTrain)    # train predict model

print("train score:", lm.score(XTrain, yTrain))
print("test score:", lm.score(XTest, yTest))
predict_value = 0.7
predict = lm.predict([[predict_value]])
print(predict)

--- volatile acidity ---
train score: 0.1656069887392485
test score: 0.12073721055673203
[5.30786745]


----------------------------------------------------
- citric acid

In [55]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

i = 2
print("---", df.columns[i], "---")
X = pd.DataFrame(x[i], columns=[str(df.columns[i])])
y = pd.Series(y)
XTrain, XTest, yTrain, yTest = train_test_split(X, y, test_size=0.3,
                                                random_state=5)
lm = LinearRegression()
lm.fit(XTrain, yTrain)    # train predict model

print("train score:", lm.score(XTrain, yTrain))
print("test score:", lm.score(XTest, yTest))
predict_value = 0
predict = lm.predict([[predict_value]])
print(predict)

--- citric acid ---
train score: 0.06561935578154465
test score: 0.015253671900837484
[5.34167231]


----------------------------------------------------
- sulphates

In [53]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

i = 9
print("---", df.columns[i], "---")
X = pd.DataFrame(x[i], columns=[str(df.columns[i])])
y = pd.Series(y)
XTrain, XTest, yTrain, yTest = train_test_split(X, y, test_size=0.3,
                                                random_state=5)
lm = LinearRegression()
lm.fit(XTrain, yTrain)    # train predict model

print("train score:", lm.score(XTrain, yTrain))
print("test score:", lm.score(XTest, yTest))
predict_value = 0.56
predict = lm.predict([[predict_value]])
print(predict)

--- sulphates ---
train score: 0.05501311147804444
test score: 0.077603685731918
[5.51828606]


----------------------------------------------------
- alcohol

In [54]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

i = 10
print("---", df.columns[i], "---")
X = pd.DataFrame(x[i], columns=[str(df.columns[i])])
y = pd.Series(y)
XTrain, XTest, yTrain, yTest = train_test_split(X, y, test_size=0.3,
                                                random_state=5)
lm = LinearRegression()
lm.fit(XTrain, yTrain)    # train predict model

print("train score:", lm.score(XTrain, yTrain))
print("test score:", lm.score(XTest, yTest))
predict_value = 9.4
predict = lm.predict([[predict_value]])
print(predict)

--- alcohol ---
train score: 0.21218726592524262
test score: 0.2579772813167879
[5.26796053]


#### 3. 結論
對volatile acidity、citric acid、sulphates、alcohol 分別測試後，  
分數高低依序為: alcohol > volatile acidity > sulphates > citric acid。  
且會對應於四者關聯係數之高低: |0.48| > |-0.39| > |0.25| > |0.23|，  

實際操作數據對四者進行預測，正確值為 5，  
四者分別得出的數據為: 5.26796053、5.30786745、5.34167231、5.51828606，  
符合前面預測出的成果，其中以 alcohol 的誤差最小，  
因此最終得出結論: 使用 alcohol 進行預測，得出來的效果會最好。
